### ISSUE TITLE
[Question]: Does it work on CI?

### ISSUE THREAD

**Issue by hffmnn** [LukeMathWalker/cargo-chef#112]
I understand, that I could use this to speed up local builds with Docker, but is it be possible to use this on CI also? What would one need to cache, so that the `cargo-chef` steps get cached?

**Comment by LukeMathWalker** [comment]
This is not meant for CI usage - you are better off caching the `target` folder directly in a CI environment.

**Comment by hffmnn** [comment]
Ok, thanks a lot for the clarification!

**Comment by alex35mil** [comment]
@LukeMathWalker Sorry for bumping the old issue, but is there a way to handle deployments from CI? I need to build a docker image to push it to the registry, but it doesn't seem like there's a straightforward way to cache the layer with dependencies. I tried to build a base image (the stage where `cargo chef cook` is called) and cache it, but it doesn't work (guessing) due to `COPY . .` on the planner stage. So, the cache is invalidated, and the cached image is not being used.

**Comment by alex35mil** [comment]
I have an idea. Instead of having a single Dockerfile, I can split it in two: one builds deps, and the second one builds a binary. On CI, I hash Cargo.lock (maybe something else) and check if the base image with this tag exists in the registry. If it’s there, I use it as a base image for the binary image and skip building the deps image. If the deps image with the tag doesn’t exist, I build it, tag it, and push it. Then, I build the binary image off of it. Seems like it should work.

**Comment by LukeMathWalker** [comment]
It sounds like you have an issue with Docker layer caching in your CI setup.
If it works locally, it should work in CI _if_ layers are cached correctly by your setup.

**Comment by alex35mil** [comment]
Maybe I don't cache what needs to be cached? What I do is build an image of the stage where I build deps and then cache it. That's it. Here are my `Dockerfile` and part of CI config (Azure Pipelines).

<details>
<summary>Dockerfile</summary>

```dockerfile
ARG BASE_IMAGE=deps-builder

FROM lukemathwalker/cargo-chef:latest-rust-1.74.1 AS chef
WORKDIR /server

FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

FROM chef AS deps-builder
COPY --from=planner /server/recipe.json recipe.json
RUN cargo chef cook --release --recipe-path recipe.json

FROM ${BASE_IMAGE} AS app-builder
COPY . .
RUN cargo build --release --bin server

FROM debian:bookworm-slim AS runtime
WORKDIR /
COPY --from=app-builder /server/target/release/server /usr/local/bin
ENTRYPOINT ["/usr/local/bin/server"]
```
</details>

<details>
</details>

**Comment by alex35mil** [comment]
Maybe I don't cache what needs to be cached? What I do is build an image of the stage where I build deps and then cache it. That's it. Here are my `Dockerfile` and part of CI config (Azure Pipelines).

When I print images list after the cache restoration (before the base build) and after the base build, image id of the base image is changed.

<details>
<summary>Dockerfile</summary>

```dockerfile
ARG BASE_IMAGE=deps-builder

FROM lukemathwalker/cargo-chef:latest-rust-1.74.1 AS chef
WORKDIR /server

FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

FROM chef AS deps-builder
COPY --from=planner /server/recipe.json recipe.json
RUN cargo chef cook --release --recipe-path recipe.json

FROM ${BASE_IMAGE} AS app-builder
COPY . .
RUN cargo build --release --bin server

FROM debian:bookworm-slim AS runtime
WORKDIR /
COPY --from=app-builder /server/target/release/server /usr/local/bin
ENTRYPOINT ["/usr/local/bin/server"]
```
</details>

<details>
<summary>CI config</summary>

```yaml
      - script: |
          DEPS_HASH=$(md5sum Cargo.lock | awk '{ print $1 }')
          echo "##vso[task.setvariable variable=CARGO_DEPS_HASH]$DEPS_HASH"
          echo "##vso[task.setvariable variable=DOCKER_REPOSITORY]$(...)/server"
          echo "##vso[task.setvariable variable=DOCKER_IMAGE_TAG]$(Build.SourceVersion)"
          echo "##vso[task.setvariable variable=DOCKER_FILE_PATH]./Dockerfile"
        displayName: "Set variables"

      - task: Cache@2
        displayName: "Setup docker cache"
        inputs:
          key: '"docker" | "$(Agent.OS)" | "$(CARGO_DEPS_HASH)" | "v1"'
          path: $(Pipeline.Workspace)/docker
          cacheHitVar: CACHE_RESTORED

      - script: docker load -i $(Pipeline.Workspace)/docker/cache.tar
        displayName: Restore Docker cache
        condition: and(not(canceled()), eq(variables.CACHE_RESTORED, 'true'))

      - script: docker image ls
        displayName: Print docker images after cache restoration

      - task: Docker@2
        displayName: "Build base Docker image"
        inputs:
          command: build
          arguments: --target deps-builder
          repository: $(DOCKER_REPOSITORY)-base
          dockerfile: $(DOCKER_FILE_PATH)
          tags: |
            $(CARGO_DEPS_HASH)

      - script: docker image ls
        displayName: Print docker images after the base build

      - script: |
          mkdir -p $(Pipeline.Workspace)/docker
          docker save -o $(Pipeline.Workspace)/docker/cache.tar $(DOCKER_REPOSITORY)-base:$(CARGO_DEPS_HASH)
        displayName: "Save Docker cache"
        condition: and(not(canceled()), not(failed()), ne(variables.CACHE_RESTORED, 'true'))

      - task: Docker@2
        displayName: "Build final Docker image"
        inputs:
          command: build
          arguments: --build-arg BASE_IMAGE=$(DOCKER_REPOSITORY)-base:$(CARGO_DEPS_HASH)
          repository: $(DOCKER_REPOSITORY)
          dockerfile: $(DOCKER_FILE_PATH)
          tags: |
            $(DOCKER_IMAGE_TAG)

```
</details>

**Comment by LukeMathWalker** [comment]
I'm sorry, but I'm not going to debug your pipelines 😅 Your Dockerfile is correct, as you confirmed since caching works as expected locally.
I'm not familiar with how Azure Pipelines work nor how they do layer caching.

**Comment by alex35mil** [comment]
By "working locally" I meant it successfully caches deps if I use it locally as suggested in the README, not the way I tried to use it on CI. I was able to make it work as I described in the follow-up comment by splitting the dockerfile into multiple and caching the image with dependencies.

Another quick question: I inspected `recipe.json`, and it seems to be agnostic to where I run `cargo chef prepare`: host vs docker. Is this correct? The context is that I use `recipe.json` instead of `Cargo.lock` to get the build fingerprint. I want to run it on the host in the first CI job, where I test the build, and then pass the generated on the host `recipe.json` to the deployment job, where I build the release in a docker, so I could skip the step where I generate the `recipe.json` inside the docker. Just want to make sure it is a safe thing to do.
